lib/srv/db/mysql/engine.go

```diff
@@ -191,6 +204,21 @@ func (e *Engine) connect(ctx context.Context, sessionCtx *common.Session) (*clie
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
+		// workaround issue generating ephemeral certs for secure connections
+		// by creating a TLS connection to the cloud proxy port and
+		// overridding mysql client's connection. mysql on the default port
+		// does not trust the ephemeral cert's CA but cloud proxy does.
+		if len(tlsConfig.Certificates) > 0 { // has client certificate
```

Comment:
TBH it feels a little strange that we're looking at client certificates in the config to determine whether we need to connect to the proxy. Wouldn't the following logic be more straightforward:
1. We fetch the GCP instance here.
2. If it has "require ssl" flag set, we fetch ephemeral cert, add it to tlsConfig and update the address to point to proxy.

This can be moved in to a helper method on the engine as to not to crowd the connect code. WDYT?

Reply:
I thought this was a bit strange as well. It stemmed from trying to find a central location to generate the ephemeral cert for both postgres and mysql; which is why it ended up in common/getTLSConfigVerifyFull. But since the connect methods in both postgres and mysql have everything needed for the ephemeral cert, there is nothing stopping us from making it an exported method in common and calling it from the postgres and mysql connect methods. I'll make the change.

Reply:
This stood out to me as well and was what I was thinking about when I mentioned it would be nice to have the GCP instance available for downstream code (so this part could check `RequireSsl` as well).

I noticed the engines access cloud clients and tls config code through the `Auth` interface, so I was trying to extend the `dbAuth` implementation by placing all tls config code in the same location (`getTLSConfigVerifyFull`). It did feel strange to make API calls there, but seemed nice to have all the config logic in the same place.

For now, to follow the same structure, I can just add `GetGCPDatabaseInstance` and `GenerateGCPEphemeralCert` functions to the `Auth` interface and let the engines each make calls from their connect code like the other token calls. We'll have to repeat some code in both engines, but it won't be much.

(For future consideration) One thing nice about the `Auth` interface is that it simplifies cloud client APIs and only exposes methods we need. The problem with `Auth` is that it is designed specifically for the engines. I'd have to think about this a bit more, but maybe it would be nice to create a cloud client interface specific to our needs a bit higher up the stack. We could then solve a couple problems at the same time. We could only expose methods from cloud clients we care about and tailor the method signatures to be more friendly (GCP call flow is silly due to them generating code). And  we could create and return our own data structures; e.g. having to check for `nil` fields in the `DatabaseInstance` code everywhere it is used is nonsense compared to doing it once and setting a `RequireSsl` field in our own struct. Perhaps the issue here is `Auth` should really be for creating teleport tls configs and there should be another interface for cloud clients.

@r0mant

Reply:
@jimbishopp Why don't we just pass CloudClients to MySQL engine and use it to get the GCP instance here?

Reply:
@r0mant  I started to do that last night but it seemed strange to access cloud clients two different ways (Auth and CloudClients).

Reply:
No reason :) I think it's fine, I wouldn't overthink it. I think it's perfectly reasonable to have access to cloud APIs in the engine and not being constrained by Auth interface.